File: bindings/zig/src/event.zig
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

const std = @import("std");
const c = @import("c.zig");
const types = @import("types.zig");
const util = @import("util.zig");

pub fn poll(node: types.Node, timeout_ms: c_int, app_data_buf: []u8) types.Error!types.Event {
    var event: types.Event = std.mem.zeroes(types.Event);
    if (app_data_buf.len > 0) {
        event.app_data = app_data_buf.ptr;
        event.app_data_cap = app_data_buf.len;
    }
    try types.mapCode(c.rns_event_poll(types.asU64(node), &event, timeout_ms));
    return event;
}

pub fn setCallback(node: types.Node, callback: ?types.EventCallback, user_data: ?*anyopaque) types.Error!void {
    try types.mapCode(c.rns_set_event_callback(types.asU64(node), callback, user_data));
}

pub fn appData(event: *const types.Event) []u8 {
    if (event.app_data == null or event.app_data_len == 0) return &.{};
    return event.app_data.?[0..event.app_data_len];
}

pub fn path(event: *const types.Event) []const u8 {
    return util.cstringField(&event.path);
}

pub fn errorMessage(event: *const types.Event) []const u8 {
    return util.cstringField(&event.error_message);
}

pub fn linkId(event: *const types.Event) []const u8 {
    return event.link_id[0..event.link_id_len];
}

pub fn destinationHash(event: *const types.Event) []const u8 {
    return event.destination_hash[0..event.destination_hash_len];
}

pub fn identityHash(event: *const types.Event) []const u8 {
    return event.identity_hash[0..event.identity_hash_len];
}

pub fn requestId(event: *const types.Event) []const u8 {
    return event.request_id[0..event.request_id_len];
}

File: bindings/rust/src/event.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

use crate::error::{map_code, Result};
use crate::ffi::{self, EventKind, HASH_LEN};
use crate::node::Node;
use crate::util;

pub struct Event {
    raw: ffi::Event,
    app_data: Vec<u8>,
    path: String,
    error_message: String,
}

impl Event {
    pub fn poll(node: &Node, timeout_ms: i32, app_data_buf: &mut [u8]) -> Result<Self> {
        let mut raw = unsafe { std::mem::zeroed::<ffi::Event>() };
        if !app_data_buf.is_empty() {
            raw.app_data = app_data_buf.as_mut_ptr();
            raw.app_data_cap = app_data_buf.len();
        }
        map_code(unsafe { ffi::rns_event_poll(node.handle(), &mut raw, timeout_ms) })?;
        let app_data = if raw.app_data.is_null() || raw.app_data_len == 0 {
            Vec::new()
        } else {
            unsafe { std::slice::from_raw_parts(raw.app_data, raw.app_data_len).to_vec() }
        };
        Ok(Self {
            path: util::cstr_field(&raw.path),
            error_message: util::cstr_field(&raw.error_message),
            raw,
            app_data,
        })
    }

    pub fn kind(&self) -> EventKind {
        match self.raw.kind {
            0 => EventKind::None,
            1 => EventKind::Announce,
            2 => EventKind::LinkEstablished,
            3 => EventKind::LinkFailed,
            4 => EventKind::LinkData,
            5 => EventKind::LinkClosed,
            6 => EventKind::RequestIncoming,
            7 => EventKind::RequestResponse,
            8 => EventKind::RequestFailed,
            9 => EventKind::ResourceStarted,
            10 => EventKind::ResourceConcluded,
            11 => EventKind::DestinationData,
            _ => EventKind::None,
        }
    }

    pub fn hops(&self) -> u8 {
        self.raw.hops
    }

    pub fn link_id(&self) -> &[u8] {
        &self.raw.link_id[..self.raw.link_id_len]
    }

    pub fn destination_hash(&self) -> &[u8] {
        &self.raw.destination_hash[..self.raw.destination_hash_len]
    }

    pub fn identity_hash(&self) -> &[u8] {
        &self.raw.identity_hash[..self.raw.identity_hash_len]
    }

    pub fn request_id(&self) -> &[u8] {
        &self.raw.request_id[..self.raw.request_id_len]
    }

    pub fn path(&self) -> &str {
        &self.path
    }

    pub fn error_message(&self) -> &str {
        &self.error_message
    }

    pub fn app_data(&self) -> &[u8] {
        &self.app_data
    }

    pub fn path_truncated(&self) -> bool {
        self.raw.path_truncated != 0
    }

    pub fn error_message_truncated(&self) -> bool {
        self.raw.error_message_truncated != 0
    }

    pub fn app_data_truncated(&self) -> bool {
        self.raw.app_data_truncated != 0
    }
}

pub fn hash_eq(a: &[u8], b: &[u8; HASH_LEN]) -> bool {
    a.len() == HASH_LEN && a == b.as_slice()
}
